README.markdown
README
======


Summary
-------

ISO::IBAN implements the IBAN (International Bank Account Number) specification as per ISO 13616-1.
It provides methods to generate valid IBAN numbers from components, or to validate a given IBAN.


Installation
------------

### Via rubygems

    gem install iso-iban

### From github

    git clone https://github.com/apeiros/iso-iban.git
    cd iso-iban
    rm -r *.gem
    gem build *.gemspec
    gem install *.gem


Usage
-----

    require 'iso/iban'
    ISO::IBAN.valid?('CH35 1234 5987 6543 2109 A')       # => true
    ISO::IBAN.validate('CH37 1234 5987 6543 2109 A')     # => [:invalid_checksum]
    ISO::IBAN.generate('CH', '12345', '987')             # => #<ISO::IBAN CH76 1234 5000 0000 0098 7>
    iban = ISO::IBAN.parse('CH35 1234 5987 6543 2109 A') # => #<ISO::IBAN CH35 1234 5987 6543 2109 A>
    iban = ISO::IBAN.new('CH351234598765432109A')        # => #<ISO::IBAN CH35 1234 5987 6543 2109 A>
    iban.formatted       # => "CH35 1234 5987 6543 2109 A"
    iban.compact         # => "CH351234598765432109A"
    iban.country         # => "CH"
    iban.checksum_digits # => "35"
    iban.bank_code       # => "12345"
    iban.account_code    # => "98765432109A"
    iban.valid?          # => true
    iban.validate        # => []

**Note:** iso/iban automatically loads the IBAN specifications delivered with the gem. If you do not wish
those to be loaded, `require 'iso/iban/no_autoload'` instead.


ENV
---

ISO::IBAN.load_specifications (which is automatically called when you require 'iso/iban') uses the
ENV variable `IBAN_SPECIFICATIONS` to determine where to look for IBAN specifications. If that
variable is not set, it will default to the datafile delivered with the gem.


Links
-----

* [Online API Documentation](http://rdoc.info/github/apeiros/iso-iban/)
* [Public Repository](https://github.com/apeiros/iso-iban)
* [Bug Reporting](https://github.com/apeiros/iso-iban/issues)
* [RubyGems Site](https://rubygems.org/gems/iso-iban)


Contributors
------------

* Carsten Wirth (ISO::IBAN#parse accepts nil)
* John Cant (Travis CI)


License
-------

You can use this code under the {file:LICENSE.txt BSD-2-Clause License}, free of charge.
If you need a different license, please ask the author.


NEWS.markdown
NEWS
====

0.1.3
-----

* Accept lower case IBANs and convert them to uppercase.
* ISO::IBAN::parse (and in consequence ::parse!, ::valid? and ::validate) accept nil as argument.
* Added section "Contributors" to README.
* Added NEWS.markdown.


0.1.2
-----

* Fix for bug which caused templated IBANs not to validate properly.


0.1.1
-----

* Fix for bug which caused ISO::IBAN::valid? and family to raise an exception on invalid characters.


0.1.0
-----

* First tagged release.


test/runner.rb
# run with `ruby test/runner.rb`
# if you only want to run a single test-file: `ruby test/runner.rb testfile.rb`

if ENV['COVERAGE']
  require 'simplecov'
  SimpleCov.start
end

$LOAD_PATH << File.expand_path('../../lib', __FILE__)
$LOAD_PATH << File.expand_path('../../test/lib', __FILE__)
TEST_DIR    = File.expand_path('../../test', __FILE__)

require 'test/unit'
require 'helper'

units = (ARGV.empty? || ENV['RAKE']) ? Dir["#{TEST_DIR}/unit/**/*.rb"] : ARGV

units.each do |unit|
  puts unit
  load unit
end


test/lib/helper.rb
# encoding: utf-8

require 'stringio'
require 'yaml'
require 'test/unit/assertions'
require 'fileutils'

module Kernel
  module_function
  def reset_test_files
    data_dir = "#{TEST_DIR}/data"
    tmp_dir  = "#{TEST_DIR}/tmp"
    FileUtils.rm_r(tmp_dir) if File.directory?(tmp_dir)
    FileUtils.cp_r(data_dir, tmp_dir)
  end

  def test_file(name)
    "#{TEST_DIR}/tmp/#{name}"
  end
end


module TestSuite
  attr_accessor :name
end

module Kernel
  def suite(name, &block)
    klass = Class.new(Test::Unit::TestCase)
    klass.extend TestSuite
    klass.name = "Suite #{name}"
    klass.class_eval(&block)

    klass
  end
  module_function :suite
end

class Test::Unit::TestCase
  def self.inherited(by)
    by.init
    super
  end

  def self.init
    @setups = []
  end

  def self.setup(&block)
    @setups ||= []
    @setups << block
  end

  class << self
    attr_reader :setups
  end

  def setup
    self.class.setups.each do |setup|
      instance_eval(&setup)
    end
    super
  end

  def self.suite(name, &block)
    klass = Class.new(Test::Unit::TestCase)
    klass.extend TestSuite
    klass.name = "#{self.name} #{name}"
    klass.class_eval(&block)

    klass
  end

  def self.test(desc, &impl)
    define_method("test #{desc}", &impl)
  end

  def capture_stdout
    captured  = StringIO.new
    $stdout   = captured
    yield
    captured.string
  ensure
    $stdout = STDOUT
  end
end


test/unit/lib/iso/iban.rb
# encoding: utf-8

require 'iso/iban/no_autoload'

suite "ISO::IBAN" do
  setup do
    ISO::IBAN.instance_variable_set(:@specifications, {'CH' => ISO::IBAN::Specification.new("Switzerland", "CH", "CH2!n5!n12!c", 21, "5!n12!c", 17, 4, 8, nil, nil)})
  end

  # since the specs provided by swift are a mess, we do a sanity check
  test 'Specifications provided by SWIFT' do
    ISO::IBAN.load_specifications
    ISO::IBAN.specifications.each do |country, spec|
      structure_length = spec.iban_structure.scan(/([A-Z]+)|(\d+)(!?)([nac])/).map { |exact, length, fixed, code|
        exact ? exact.length : length.to_i
      }.inject(:+)

      assert_equal spec.iban_length, structure_length, "Spec for country #{country} is invalid, invalid length"
      assert_equal spec.bban_structure, spec.iban_structure[5..-1], "Spec for country #{country} is invalid, iban differs from country + 2!n + bban"
      assert_equal spec.bban_length, spec.iban_length-4, "Spec for country #{country} is invalid, iban length differs from bban length + 4"
    end
  end

  test 'Random IBAN generation' do
    ISO::IBAN.load_specifications
    ISO::IBAN.specifications.each do |country, spec|
      assert ISO::IBAN.random(country).valid?, "Random IBAN generator for country #{country} is invalid"
    end
  end

  test 'ISO::IBAN::generate problem, TODO (this test is expected to fail at the moment)' do
    ISO::IBAN.instance_variable_set(:@specifications, {'BG' => ISO::IBAN::Specification.new("Bulgaria", "BG", "BG2!n4!a4!n2!n8!c", 22, "4!a4!n2!n8!c", 18, 4, 7, 8, 11)})
    assert ISO::IBAN.generate('BG', 'AAAA', '2', 'C').valid? # this works now
    assert ISO::IBAN.generate('BG', 'A', '2', 'C').valid? # this still fails, because ISO::IBAN::generate can't pad 'a' format fields
  end

  test 'ISO::IBAN::load_specifications' do
    reset_test_files

    ISO::IBAN.load_specifications(test_file('test_spec.yaml'))
    specs = ISO::IBAN.specifications
    assert_kind_of Hash, specs
    specs.each do |key, value|
      assert_kind_of String, key
      assert_kind_of ISO::IBAN::Specification, value
      assert_equal key, value.a2_country_code
    end
    assert_equal ["AL", "AD"], specs.keys
  end

  test 'ISO::IBAN::numerify' do
    assert_equal 121735123459876543210910, ISO::IBAN.numerify('CH351234598765432109A')
    assert_raise ArgumentError do
      ISO::IBAN.numerify('CH99 1234 5987 6543 2109 A')
    end
  end

  test 'ISO::IBAN::generate' do
    assert_equal 'CH76 0012 3000 0000 9876 B', ISO::IBAN.generate('CH', '123', '9876B').formatted
  end

  test 'ISO::IBAN::valid?' do
    assert ISO::IBAN.valid?('CH35 1234 5987 6543 2109 A')
    assert !ISO::IBAN.valid?('CH99 1234 5987 6543 2109 A')
    assert !ISO::IBAN.valid?('CH150+') # raised before 0.1.0, fixed in 0.1.1
    assert !ISO::IBAN.valid?('foo')
    assert_equal false, ISO::IBAN.valid?(nil)
    assert_equal false, ISO::IBAN.valid?('')
  end

  test 'ISO::IBAN::validate' do
    assert_equal [], ISO::IBAN.validate('CH35 1234 5987 6543 2109 A')
    assert_equal [:invalid_checksum], ISO::IBAN.validate('CH99 1234 5987 6543 2109 A')
    assert_equal [:invalid_format], ISO::IBAN.validate('CH86 X234 5987 6543 2109 A')
    assert_equal [:invalid_checksum, :invalid_format], ISO::IBAN.validate('CH99 X234 5987 6543 2109 A')
    assert_equal [:invalid_length, :invalid_format], ISO::IBAN.validate('CH83 X234 5987 6543 2109 AB')
    assert_equal [:invalid_checksum, :invalid_length, :invalid_format], ISO::IBAN.validate('CH99 X234 5987 6543 2109 AB')
    assert_equal [:invalid_country, :invalid_checksum, :invalid_length, :invalid_format], ISO::IBAN.validate('XX35 1234 5987 6543 2109 A')
    assert_equal [:invalid_characters, :invalid_checksum, :invalid_length, :invalid_format], ISO::IBAN.validate('CH35 1234 5987 6543 2109 Ä')
    assert_equal [:invalid_characters, :invalid_country, :invalid_checksum, :invalid_length, :invalid_format], ISO::IBAN.validate('XX35 1234 5987 6543 2109 Ä')
    assert_equal [:invalid_characters, :invalid_country,  :invalid_checksum, :invalid_length, :invalid_format], ISO::IBAN.validate(nil)
    assert_equal [:invalid_characters, :invalid_country,  :invalid_checksum, :invalid_length, :invalid_format], ISO::IBAN.validate('')
  end

  test "ISO::IBAN::parse" do
    assert_kind_of ISO::IBAN, ISO::IBAN.parse('CH35 1234 5987 6543 2109 A')
    assert_equal 'CH351234598765432109A', ISO::IBAN.parse('CH35 1234 5987 6543 2109 A').compact
  end

  test "ISO::IBAN::new" do
    assert_kind_of ISO::IBAN, ISO::IBAN.new('CH351234598765432109A')
  end

  test "ISO::IBAN#formatted" do
    assert_equal 'CH35 1234 5987 6543 2109 A', ISO::IBAN.parse('CH35 1234 5987 6543 2109 A').formatted
    assert_equal 'CH35 1234 5987 6543 2109 A', ISO::IBAN.parse('CH351234598765432109A').formatted
  end

  test "ISO::IBAN#compact" do
    assert_equal 'CH351234598765432109A', ISO::IBAN.parse('CH35 1234 5987 6543 2109 A').compact
    assert_equal 'CH351234598765432109A', ISO::IBAN.parse('CH351234598765432109A').compact
  end

  test "ISO::IBAN#to_s" do
    assert_equal 'CH351234598765432109A', ISO::IBAN.parse('CH35 1234 5987 6543 2109 A').to_s
    assert_equal 'CH351234598765432109A', ISO::IBAN.parse('CH351234598765432109A').to_s
  end

  test "ISO::IBAN#country" do
    assert_equal 'CH', ISO::IBAN.parse('CH35 1234 5987 6543 2109 A').country
  end

  test "ISO::IBAN#checksum_digits" do
    assert_equal '35', ISO::IBAN.parse('CH35 1234 5987 6543 2109 A').checksum_digits
  end

  test "ISO::IBAN#bban" do
    assert_equal '1234598765432109A', ISO::IBAN.parse('CH35 1234 5987 6543 2109 A').bban
  end

  test "ISO::IBAN#bank_code" do
    assert_equal '12345', ISO::IBAN.parse('CH35 1234 5987 6543 2109 A').bank_code
  end

  test "ISO::IBAN#branch_code" do
    ISO::IBAN.instance_variable_set(:@specifications, {'BG' => ISO::IBAN::Specification.new("Bulgaria", "BG", "BG2!n4!a4!n2!n8!c", 22, "4!a4!n2!n8!c", 18, 4, 7, 8, 11)})
    assert_equal "0002", ISO::IBAN.parse('BG69 0001 0002 0300 0000 04').branch_code
  end

  test "ISO::IBAN#account_code" do
    assert_equal '98765432109A', ISO::IBAN.parse('CH35 1234 5987 6543 2109 A').account_code
  end

  test "ISO::IBAN#valid?" do
    assert_equal true,  ISO::IBAN.parse('CH35 1234 5987 6543 2109 A').valid?
    assert_equal false, ISO::IBAN.parse('CH99 1234 5987 6543 2109 A').valid?
  end

  test "ISO::IBAN#validate" do
    assert_equal [], ISO::IBAN.parse('CH35 1234 5987 6543 2109 A').validate
  end

  test 'ISO::IBAN#invalid_characters' do
    assert_equal ["\xC3".b, "\x84".b], ISO::IBAN.parse('Ä').invalid_characters
    assert_equal ['Ä'],                ISO::IBAN.parse('Ä').invalid_characters('utf-8')
  end

  test "ISO::IBAN#<=>" do
    iban0 = ISO::IBAN.generate('CH', '0', '0')
    iban1 = ISO::IBAN.generate('CH', '0', '97') # 97 to have the same checksum
    assert_equal( -1, iban0 <=> iban1         )
    assert_equal(  0, iban0 <=> iban0         )
    assert_equal(  1, iban1 <=> iban0         )
    assert_equal(nil, iban0 <=> "incomparable")
    assert_equal(nil, "incomparable" <=> iban0)
  end

  test "ISO::IBAN#to_a" do
    assert_equal %w[CH 35 12345 98765432109A], ISO::IBAN.parse("CH351234598765432109A").to_a
    assert_equal [], ISO::IBAN.parse("XX351234598765432109A").to_a # no specification
  end

  test "ISO::IBAN#inspect" do
    assert_equal "#<ISO::IBAN CH35 1234 5987 6543 2109 A>", ISO::IBAN.parse('CH35 1234 5987 6543 2109 A').inspect
  end
end


lib/iso/iban/specification.rb
# encoding: utf-8

module ISO
  class IBAN

    # Specification of the IBAN format for one country. Every country has its
    # own specification of the IBAN format.
    # SWIFT is the official authority where those formats are registered.
    class Specification

      # A mapping from SWIFT structure specification to PCRE regex.
      StructureCodes = {
        'n' => '\d',
        'a' => '[A-Z]',
        'c' => '[A-Za-z0-9]',
        'e' => ' ',
      }

      # Load the specifications YAML.
      #
      # @return [Hash<String => ISO::IBAN::Specification>]
      def self.load_yaml(path)
        Hash[YAML.load_file(path).map { |country, spec| [country, new(*spec)] }]
      end

      # Parse the SWIFT provided file (which sadly is a huge mess and not machine friendly at all).
      #
      # @return [Array<ISO::IBAN::Specification>] an array with all specifications.
      def self.parse_file(path)
        File.read(path, encoding: Encoding::Windows_1252).encode(Encoding::UTF_8).split("\r\n").tap(&:shift).flat_map { |line|
          country_name, country_codes, iban_structure_raw, iban_length, bban_structure, bban_length, bank_position = line.split(/\t/).values_at(0,1,11,12,4,5,6)
          codes        = country_codes.size == 2 ? [country_codes] : country_codes.scan(/\b[A-Z]{2}\b/)
          primary_code = codes.first
          bank_position_from, bank_position_to, branch_position_from, branch_position_to = bank_position.match(/(?:[Pp]ositions?|) (\d+)-(\d+)(?:.*Branch identifier positions?: (\d+)-(\d+))?/).captures.map { |pos| pos && pos.to_i+3 }

          codes.map { |a2_country_code|
            iban_structure = iban_structure_raw[/#{a2_country_code}[acen\!\d]*/] || iban_structure_raw[/#{primary_code}[acen\!\d]*/]
            bban_structure = bban_structure[/[acen\!\d]*/]

            new(
              country_name.strip,
              a2_country_code,
              iban_structure,
              iban_length.to_i,
              bban_structure.strip,
              bban_length.to_i,
              bank_position_from,
              bank_position_to,
              branch_position_from,
              branch_position_to
            )
          }
        }
      end

      # *n:   Digits (numeric characters 0 to 9 only)
      # *a:   Upper case letters (alphabetic characters A-Z only)
      # *c:   upper and lower case alphanumeric characters (A-Z, a-z and 0-9)
      # *e:   blank space
      # *nn!: fixed length
      # *nn:  maximum length
      #
      # Example: "AL2!n8!n16!c"
      def self.structure_regex(structure, anchored=true)
        source = structure.scan(/([A-Z]+)|(\d+)(!?)([nac])/).map { |exact, length, fixed, code|
          if exact
            Regexp.escape(exact)
          else
            StructureCodes[code]+(fixed ? "{#{length}}" : "{,#{length}}")
          end
        }.join(')(')

        anchored ? /\A(#{source})\z/ : /(#{source})/n
      end

      attr_reader :country_name,
                  :a2_country_code,
                  :iban_structure,
                  :iban_length,
                  :bban_structure,
                  :bban_length,
                  :bank_position_from,
                  :bank_position_to,
                  :branch_position_from,
                  :branch_position_to

      def initialize(country_name, a2_country_code, iban_structure, iban_length, bban_structure, bban_length, bank_position_from, bank_position_to, branch_position_from, branch_position_to)
        @country_name         = country_name
        @a2_country_code      = a2_country_code
        @iban_structure       = iban_structure
        @iban_length          = iban_length
        @bban_structure       = bban_structure
        @bban_length          = bban_length
        @bank_position_from   = bank_position_from
        @bank_position_to     = bank_position_to
        @branch_position_from = branch_position_from
        @branch_position_to   = branch_position_to
      end

      # @return [Regexp] A regex to verify the structure of the IBAN.
      def iban_regex
        @_iban_regex ||= self.class.structure_regex(@iban_structure)
      end

      # @return [Regexp] A regex to identify the structure of the IBAN, without anchors.
      def unanchored_iban_regex
        self.class.structure_regex(@iban_structure, false)
      end

      # @return [Array<Integer>] An array with the lengths of all components.
      def component_lengths
        [bank_code_length, branch_code_length, account_code_length].tap { |lengths| lengths.delete(0) }
      end

      # @return [Fixnum]
      #   The length of the bank code in the IBAN, 0 if the IBAN has no bank code.
      def bank_code_length
        @bank_position_from && @bank_position_to ? @bank_position_to-@bank_position_from+1 : 0
      end

      # @return [Fixnum]
      #   The length of the bank code in the IBAN, 0 if the IBAN has no branch code.
      def branch_code_length
        @branch_position_from && @branch_position_to ? @branch_position_to-@branch_position_from+1 : 0
      end

      # @return [Fixnum]
      #   The length of the account code in the IBAN.
      def account_code_length
        bban_length-bank_code_length-branch_code_length
      end

      # @return [Array] An array with the Specification properties. Used for serialization.
      def to_a
        [
          @country_name,
          @a2_country_code,
          @iban_structure,
          @iban_length,
          @bban_structure,
          @bban_length,
          @bank_position_from,
          @bank_position_to,
          @branch_position_from,
          @branch_position_to,
        ]
      end
    end
  end
end


lib/iso/iban/no_autoload.rb
# encoding: utf-8

require 'iso/iban/backports'
require 'iso/iban/invalid'
require 'iso/iban/specification'
require 'iso/iban/version'
require 'yaml'

# Container for classes which implement ISO standards.
module ISO

  # IBAN - ISO 13616-1
  #
  # Usage
  # =====
  #
  #     require 'iso/iban'
  #     ISO::IBAN.valid?('CH35 1234 5987 6543 2109 A')       # => true
  #     ISO::IBAN.validate('CH37 1234 5987 6543 2109 A')     # => [:invalid_checksum]
  #     ISO::IBAN.generate('CH', '12345', '987')             # => #<ISO::IBAN CH76 1234 5000 0000 0098 7>
  #     iban = ISO::IBAN.parse('CH35 1234 5987 6543 2109 A') # => #<ISO::IBAN CH35 1234 5987 6543 2109 A>
  #     iban = ISO::IBAN.new('CH351234598765432109A')        # => #<ISO::IBAN CH35 1234 5987 6543 2109 A>
  #     iban.formatted       # => "CH35 1234 5987 6543 2109 A"
  #     iban.compact         # => "CH351234598765432109A"
  #     iban.country         # => "CH"
  #     iban.checksum_digits # => "35"
  #     iban.bank_code       # => "12345"
  #     iban.account_code    # => "98765432109A"
  #     iban.valid?          # => true
  #     iban.validate        # => []
  #
  # General IBAN Information
  # ========================
  #
  # * **What is an IBAN?**  
  #   IBAN stands for International Bank Account Number. It is the ISO 13616
  #   international standard for numbering bank accounts. In 2006, the
  #   International Organization for Standardization (ISO) designated SWIFT as
  #   the Registration Authority for ISO 13616.
  #
  # * **Use**  
  #   The IBAN facilitates the communication and processing of cross-border
  #   transactions. It allows exchanging account identification details in a
  #   machine-readable form.
  #
  #
  # The ISO 13616 IBAN Standard
  # ===========================
  #
  # * **Structure**  
  #   The IBAN structure is defined in ISO 13616-1 and consists of a two-letter
  #   ISO 3166-1 country code, followed by two check digits and up to thirty
  #   alphanumeric characters for a BBAN (Basic Bank Account Number) which has a
  #   fixed length per country and, included within it, a bank identifier with a
  #   fixed position and a fixed length per country. The check digits are
  #   calculated based on the scheme defined in ISO/IEC 7064 (MOD97-10).
  #
  # * **Terms and definitions**  
  #     * *Bank identifier:* The identifier that uniquely identifies the financial
  #       institution and, when appropriate, the branch of that financial institution
  #       servicing an account.  
  #     * *BBAN:* basic bank account number: The identifier that uniquely identifies
  #       an individual account, at a specific financial institution, in a particular
  #       country. The BBAN includes a bank identifier of the financial institution
  #       servicing that account.  
  #     * *IBAN:* international bank account number: The expanded version of the
  #       basic bank account number (BBAN), intended for use internationally. The
  #       IBAN uniquely identifies an individual account, at a specific financial
  #       institution, in a particular country.
  #
  # * **Submitters**  
  #   Nationally-agreed, ISO13616-compliant IBAN formats are submitted to the
  #   registration authority exclusively by the National Standards Body or the
  #   National Central Bank of the country.
  class IBAN
    include Comparable

    # Character code translation used to convert an IBAN into its numeric
    # (digits-only) form
    CharacterCodes  = Hash[('0'..'9').zip('0'..'9')+('a'..'z').zip(10..35)+('A'..'Z').zip(10..35)]

    # All uppercase letters
    UpperAlpha   = [*'A'..'Z']

    # All lowercase letters
    LowerAlpha   = [*'a'..'z']

    # All digits
    Digits       = [*'0'..'9']

    # All uppercase letters, lowercase letters and digits
    AlphaNumeric = [*'A'..'Z', *'a'..'z', *'0'..'9']

    # All specifications, see ISO::IBAN::Specification
    @specifications = nil

    # @note
    #   Using `require 'iso/iban'` will automatically invoke this method.
    #   If you do not wish this behavior, `require 'iso/iban/no_autoload'` instead.
    #
    # Load the IBAN specifications file, which determines how the IBAN
    # for any given country looks.
    #
    # It will use the following sources in this order (first one which exists wins)
    #
    # * Path passed as `spec_file` parameter
    # * Path provided by the env variable `IBAN_SPECIFICATIONS`
    # * The file `../data/iso-iban/specs.yaml` relative to the lib dir
    # * The Gem datadir path
    #
    # @param [String] spec_file
    #   Override the default specifications file path.
    #
    # @return [self]
    def self.load_specifications(spec_file=nil)
      if spec_file then
        # do nothing
      elsif ENV['IBAN_SPECIFICATIONS'] then
        spec_file = ENV['IBAN_SPECIFICATIONS']
      else
        spec_file = File.expand_path('../../../../data/iso-iban/specs.yaml', __FILE__)
        if !File.file?(spec_file) && defined?(Gem) && Gem.datadir('iso-iban')
          spec_file = Gem.datadir('iso-iban')+'/specs.yaml'
        end
      end

      if spec_file && File.file?(spec_file)
        @specifications = ISO::IBAN::Specification.load_yaml(spec_file)
      elsif spec_file
        raise "Could not load IBAN specifications, specs file #{spec_file.inspect} does not exist or can't be read."
      else
        raise "Could not load IBAN specifications, no specs file found."
      end

      self
    end

    # @return [Hash<String => ISO::IBAN::Specification>]
    #   A hash with the country (ISO3166 2-letter) as key and the specification for that country as value.
    def self.specifications
      @specifications || raise("No specifications have been loaded yet - Check the docs for ISO::IBAN::load_specifications.")
    end

    # @param [String] a2_country_code
    #   The country (ISO3166 2-letter), e.g. 'CH' or 'DE'.
    #
    # @return [ISO::IBAN::Specification]
    #   The specification for the given country.
    def self.specification(a2_country_code, *default, &default_block)
      specifications.fetch(a2_country_code, *default, &default_block)
    end

    # @param [String] iban
    #   An IBAN number, either in compact or human format.
    #
    # @return [true, false]
    #   Whether the IBAN is valid.  
    #   See {#validate} for details.
    def self.valid?(iban)
      parse(iban).valid?
    end

    # @param [String] iban
    #   An IBAN number, either in compact or human format.
    #
    # @return [Array<Symbol>]
    #   An array with a code of all validation errors, empty if valid.  
    #   See {#validate} for details.
    def self.validate(iban)
      parse(iban).validate
    end

    # @param [String] iban
    #   The IBAN in either compact or human readable form.
    #
    # @return [String]
    #   The IBAN in compact form, all whitespace and dashes stripped.
    def self.strip(iban)
      iban.delete("\n\r\t -")
    end

    # Like ISO::IBAN.parse, but raises an ISO::IBAN::Invalid exception if the IBAN is invalid.
    #
    # @param [String, nil] iban_number
    #   The IBAN in either compact or human readable form.
    #
    # @return [ISO::IBAN]
    #   An IBAN instance representing the passed IBAN number.
    def self.parse!(iban_number)
      iban   = parse(iban_number)
      raise Invalid.new(iban) unless iban.valid?

      iban
    end

    # @param [String, nil] iban_number
    #   The IBAN in either compact or human readable form.
    #
    # @return [ISO::IBAN]
    #   An IBAN instance representing the passed IBAN number.
    def self.parse(iban_number)
      new(strip(iban_number || ""))
    end

    # Generate an IBAN from country code and components, automatically filling in the checksum.
    #
    # @note
    #   `generate` will pad all segments with zeros, which means it will generate invalid
    #   IBANs if you provide too short segments which are alphabetic only.  
    #   For example, `ISO::IBAN.generate('BG', 'A', '2', 'C')` generates an invalid IBAN.
    #
    # @example Generate an IBAN for UBS Switzerland with account number '12345'
    #     ISO::IBAN.generate('CH', '216', '12345') # => #<ISO::IBAN CH92 0021 6000 0000 1234 5>
    #
    # @param [String] country
    #   The ISO-3166 2-letter country code.
    #
    def self.generate(country, *components)
      spec      = specification(country)
      justified = spec.component_lengths.zip(components).map { |length, component| component.rjust(length, "0") }
      iban      = new(country+'??'+justified.join(''))
      iban.update_checksum!

      iban
    end

    # @param [String] countries
    #   A list of 2 letter country codes. If empty, all countries in ISO::IBAN::specifications are used.
    #
    # @return [ISO::IBAN] A random, valid IBAN.
    def self.random(*countries)
      countries = specifications.keys if countries.empty?
      country   = countries.sample
      account   = specification(country).iban_structure.scan(/([A-Z]+)|(\d+)(!?)([nac])/).map { |exact, length, fixed, code|
        if exact
          exact
        elsif code == 'a'
          Array.new(length.to_i) { UpperAlpha.sample }.join('')
        elsif code == 'c'
          Array.new(length.to_i) { AlphaNumeric.sample }.join('')
        elsif code == 'e'
          ' '*length.to_i
        elsif code == 'n'
          Array.new(length.to_i) { Digits.sample }.join('')
        end
      }.join('')
      account[2,2] = '??'
      iban = new(account)
      iban.update_checksum!

      iban
    end

    # Converts a String into its digits-only form, i.e. all characters a-z are replaced with their corresponding
    # digit sequences, according to the IBAN specification.
    #
    # @param [String] string
    #   The string to convert into its numeric form.
    #
    # @return [String] The string in its numeric, digits-only form.
    def self.numerify(string)
      string.downcase.gsub(/\D/) { |char|
        CharacterCodes.fetch(char) {
          raise ArgumentError, "The string contains an invalid character #{char.inspect}"
        }
      }.to_i
    end

    # @return [String] The standard form of the IBAN for machine communication, without spaces, encoded in Encoding::BINARY.
    attr_reader :compact

    # @return [String] The ISO-3166 2-letter country code (first and second character).
    attr_reader :country

    # @return [ISO::IBAN::Specification] The specification for this IBAN (determined by its country).
    attr_reader :specification

    # @param [String] iban
    #   The IBAN number, must be in compact form. Use ISO::IBAN::parse for formatted IBANs.
    def initialize(iban)
      raise ArgumentError, "String expected for iban, but got #{iban.class}" unless iban.is_a?(String)

      @compact       = iban.b.upcase
      @country       = @compact[0,2]
      @specification = self.class.specification(@country, nil)
    end

    # @example Formatted IBAN
    #
    #     ISO::IBAN.new('CH')
    #
    # @return [String] The IBAN in its formatted form, which is more human readable than the compact form.
    def formatted
      @_formatted ||= @compact.gsub(/.{4}(?=.)/, '\0 ')
    end

    # @return [Integer, nil]
    #   IBAN in its numeric form, i.e. all characters a-z are replaced with their corresponding
    #   digit sequences.  
    #   Will return nil if the IBAN is shorter than 5 characters (which makes it invalid).
    def numeric
      @compact.size < 5 ? nil : self.class.numerify(@compact[4..-1]+@compact[0,4])
    end

    # @return [true, false]
    #   Whether the IBAN is valid.
    #   See {#validate} for details.
    def valid?
      valid_characters? && valid_country? && valid_checksum? && valid_length? && valid_format?
    end

    # @note
    #   The class method {ISO::IBAN.validate} uses {ISO::IBAN.parse}, which means it will strip whitespace and
    #   dashes from the IBAN.  
    #   {ISO::IBAN#initialize} on the other hand expects the IBAN in compact format and will not strip
    #   those characters.
    #
    # Error codes:
    #
    # * :invalid_characters
    # * :invalid_country
    # * :invalid_checksum
    # * :invalid_length
    # * :invalid_format
    #
    # Invalid characters means that the IBAN contains characters which are not in the set of A-Za-z0-9. See {#invalid_characters}.  
    # Invalid country means the country is unknown (character 1 & 2 in the IBAN).  
    # Invalid checksum means the two check digits (character 3 & 4 in the IBAN).  
    # Invalid length means the IBAN does not comply with the length specified for the country of that IBAN.  
    # Invalid format means the IBAN does not comply with the format specified for the country of that IBAN.  
    #
    # @return [Array<Symbol>] An array with a code of all validation errors, empty if valid.
    def validate
      errors   = []
      errors << :invalid_characters unless valid_characters?
      errors << :invalid_country    unless valid_country?
      errors << :invalid_checksum   unless valid_characters? && valid_checksum?
      errors << :invalid_length     unless valid_length?
      errors << :invalid_format     unless valid_format?

      errors
    end

    # @return [String] The checksum digits in the IBAN (character 3 & 4).
    def checksum_digits
      @compact[2,2]
    end

    # @return [String] The BBAN of the IBAN (everything except the country code and check digits).
    def bban
      @compact[4..-1]
    end

    # @return [String, nil] The bank code part of the IBAN, nil if not applicable.
    def bank_code
      if @specification && @specification.bank_position_from && @specification.bank_position_to
        @compact[@specification.bank_position_from..@specification.bank_position_to]
      else
        nil
      end
    end

    # @return [String, nil] The branch code part of the IBAN, nil if not applicable.
    def branch_code
      if @specification && @specification.branch_position_from && @specification.branch_position_to
        @compact[@specification.branch_position_from..@specification.branch_position_to]
      else
        nil
      end
    end

    # @return [String] The account code part of the IBAN.
    def account_code
      @compact[((@specification.branch_position_to || @specification.bank_position_to || 3)+1)..-1]
    end

    # @example
    #     invalid = "hägar"
    #     invalid.encoding # => #<Encoding:UTF-8>
    #     ISO::IBAN.new(invalid).invalid_characters          # => ["\xC3", "\xA4"]
    #     ISO::IBAN.new(invalid).invalid_characters('utf-8') # => ["ä"]
    #
    # @param [String, Encoding, nil] input_encoding
    #   ISO::IBAN::new interprets the passed IBAN as binary.
    #   If you got the IBAN from a source which is not binary, you should provide that encoding.
    #   Otherwise an invalid character may be split into multiple bytes.
    #
    # @return [Array] An Array with all invalid characters.
    def invalid_characters(input_encoding=nil)
      iban = input_encoding ? @compact.dup.force_encoding(input_encoding) : @compact

      iban.gsub(/[A-Z0-9?]*/i, '').chars.to_a.uniq # to_a is for ruby <= 2.0 where String#chars returns an Enumerator
    end

    # @return [true, false] Whether IBAN consists only of valid characters.
    def valid_characters?
      @compact =~ /\A[A-Z]{2}(?:\d\d|\?\?)[A-Z0-9]*\z/in ? true : false
    end

    # @return [true, false] Whether the country of the IBAN is valid.
    def valid_country?
      @specification ? true : false
    end

    # @return [true, false] Whether the format of the IBAN is valid.
    def valid_format?
      @specification && @specification.iban_regex =~ @compact ? true : false
    end

    # @return [true, false] Whether the length of the IBAN is valid.
    def valid_length?
      @specification && @compact.size == @specification.iban_length ? true : false
    end

    # @return [true, false] Whether the checksum of the IBAN is valid.
    def valid_checksum?
      numerified = numeric()

      numerified && (numerified % 97 == 1)
    end

    # See Object#<=>
    #
    # @return [-1, 0, 1, nil]
    def <=>(other)
      other.respond_to?(:compact) ? @compact <=> other.compact : nil
    end

    # Requires that the checksum digits were left as '??', replaces them with
    # the proper checksum.
    #
    # @return [self]
    def update_checksum!
      raise "Checksum digit placeholders missing" unless @compact[2,2] == '??'

      @compact[2,2] = calculated_check_digits
      @_formatted   = nil

      self
    end

    # @return [String] The check-digits as calculated from the IBAN.
    def calculated_check_digits
      "%02d" % (98-(self.class.numerify(bban+@country)*100)%97)
    end

    # @return [true, false]
    #   Whether two ISO::IBANs are equal.
    #   Comparison is based on class and IBAN number
    def eql?(other)
      self.class.equal?(other.class) && self == other
    end

    # @return [Integer]
    #   A hash value, see Object#hash
    def hash
      [self.class, @compact].hash
    end

    # See Object#inspect
    def inspect
      sprintf "#<%p %s>", self.class, formatted
    end

    # @return [String] The compact form of the IBAN as a String.
    def to_s
      @compact.dup
    end

    # @note
    #   This method is experimental. It might change or be removed in future versions!
    #
    # @return [Array]
    #   The individual IBAN components as defined by the SWIFT specification.
    #   An empty array if this IBAN does not have a specification.
    def to_a
      @_components ||= @specification ? @compact.match(@specification.iban_regex).captures : []

      @_components.dup
    end
  end
end
